clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java

```diff
@@ -447,6 +482,19 @@ public void close() {
 
     @Override
     public void close(Duration timeout) {
+        if (timeout.toMillis() < 0)
+            throw new IllegalArgumentException("The timeout cannot be negative.");
+        try {
+            if (!closed) {
+                close(timeout, false);
+            }
+        } finally {
+            closed = true;
```

Comment:
It's interesting to set this only after the call to close completes or fails.

I could see setting it only after it completes to ensure it's really closed, or setting it with a CAS before calling the inner close so that multiple callers don't all try to close at once.

I'm not sure I see the rationale of doing it this way, though. If there is a reason I'm not seeing, could you add a comment explaining it so the code will be clear?

Reply:
actually that's a good point. I did this because it was the original implementation and I didn't put too much thinking into this.  For the current threading model, we definitely want to avoid multiple thread trying to close the client.  Can we synchronize on this method to prevent multithreading access?

Reply:
Yes you can! Although in general, you should not use synchronized methods, but instead use a synchronized block or a Lock or something.

Reply:
could you explain why closed is boolean but shouldWakeup is AtomicBoolean ?